### DISCUSSION TITLE
Additional vehicles

### DISCUSSION THREAD

**Discussion by fch-aa** [juanmanzanero/fastest-lap#6] [Q&A] [answered]
Hello, I'm interested in doing some simulations for GT cars on some local tracks. 
I managed to create the track, but I'm having a hard time figuring out what each parameter means in the vehicle definition file (and where to find them).

Do you have any pointer on how to create one for say a Porsche 911 C2S (997.1 / 991.1)?
The one I created is roughly 30secs (approx. two minutes lap time) too quick compared to IRL.

Thanks for this amazing project!

EDIT: Managed to lower the gap to ~10secs.
I think the most "wrong" parameters for my use case are the pacejka model values (c/p from the F1 model).

[porsche-997.1-c2s.txt](https://github.com/juanmanzanero/fastest-lap/files/8539686/porsche-997.1-c2s.txt)

**Comment by juanmanzanero** [comment, marked as answer]
Hello! Your project sounds very cool. You can probably dig on internet to find proper values for GT cars, I have seen some projects in the past while creating Fastest-lap.

I have not created any documentation yet since the project is still under very heavy development, although there are some people using it already. You can look for the meaning of the parameters in the references I have linked in the welcome page.

Regarding your mismodelling, I would be cool to know if you are faster in the straights (too much power) or faster in the corners (too much lateral grip from the tires). Tweaking these parameters you can probably get a better match. 

Question: what circuit are you running in?

Cheers!

**Reply by juanmanzanero** [reply]
BTW just saw in your XML file that you are using an engine power curve. This is not currently supported, only constant power (via maximum-power) works.

**Reply by fch-aa** [reply]
The track is called "La Ferté Gaucher" (LFG), it's a small track near Paris: https://www.circuitslfg.fr/
You will find attached to this message the KMZ / KML & corresponding XML file.

Understood for the power curve, I found those in the kart model XML file; this may explain the lap time difference!

There is definitively a difference in top speed, the simulation shows ~180 km/h at the end of the straight, but the real telemetry has a top speed of 168 km/h.

[lfg.zip](https://github.com/juanmanzanero/fastest-lap/files/8541326/lfg.zip)

**Reply by juanmanzanero** [reply]
Nice work! Feel free to drop a pull request with your files into the database/tracks folder. Surely you are overpowered, maybe reduce the engine maximum power, or increase the Cd

**Comment by juanmanzanero** [comment]
Hello @fch-aa , I am curious: have you arrived to anything interesting with the GT car model? Cheers

**Reply by fch-aa** [reply]
Hello @juanmanzanero, I just started looking into this but you can see attached a simple viz between two real & the simulated lap time.

Do you have any idea what should I tweak ? I would say that the simulated car seems to have way more grip in the s2 fast left hander.

You can see this part of the track here to get a better idea (not my lap): https://youtu.be/vJwfMUDNxG0?t=349

EDIT: The difference in distance is concerning, ~100m! 🤔

Cheers

![LFG_simulation_vs_real](https://user-images.githubusercontent.com/21101725/165625598-7a38b0df-e146-441e-bc58-7aa998c5590b.png)
![LFG_simulation_vs_real_time](https://user-images.githubusercontent.com/21101725/165823553-899d4752-7437-48a7-9c48-cd42b82fa443.png)
![LFG_simulation_impacted_sector](https://user-images.githubusercontent.com/21101725/165627323-3b740bfb-b5b6-41d6-9656-8e03314c0b4e.png)

**Reply by juanmanzanero** [reply]
Hello!! 

This looks very nice! It looks super super good, you have generated a kml with the trajectory?

For the issue with fast left handed, you can decrease the grip coefficient at high loads, mu-{x,y}-{front, rear}-2. Don't get too obsessed with perfectly matching a real lap, since it can happen that the real driver is a bit far from the ideal! I'd take a look at differences with the throttle/brake and so on too.

About the difference in distance, it happens that what you are seeing is not the real traveled distance, is the distance along the reference line (the one you compute with the circuit preprocessor). To get the real traveled distance you have to do something like:

`s[i] = s[i-1] + sqrt( (x[i]-x[i-1])^2 + (y[i]-y[i-1])^2 )`

I think this should explain the differences, but tell me otherwise.

Cheers!

**Reply by fch-aa** [reply]
I tried to do just that, but I'm missing something. What you see on the Google Earth screenshot is actually my real lap.
Using the following code inspired by the unit tests I can convert correctly the track boundaries, but when I try to do the same with the simulated lap "x, y" values I get a distorted path. 

I guess the Cartesian to long/lat conversion is not enough?

``` python
lon = sim_df['x'] / (R_earth * np.cos(phi0 * DEG)) * RAD + theta0
lat = sim_df['y'] / (R_earth) * RAD + phi0
```

![Simulated KML issue](https://user-images.githubusercontent.com/21101725/166138919-64d1e9de-4869-48d7-beee-0675209cc9b0.png)

The issue is most likely between the seat and the steering wheel, that's precisely what I want to see ;)

I used the notebook you created to compare F1 laps (examples/python/f1/optimal-laptime/Optimal_laptime.ipynb), in this code you did not process the distance like you mentioned above, did I miss something?

``` python
fig, ax = plt.subplots(figsize=(20,5))
ax.plot(ver_tel['Distance'], ver_tel['Speed'], color=rbr_color, label='VER')
ax.plot(ham_tel['Distance'], ham_tel['Speed'], color=mer_color, label='HAM')
ax.plot(np.array(s)-168.0,np.array(u)*3.6, label="Fastest-lap")
```

**Reply by juanmanzanero** [reply]
In the GPS plot, the casual 90º of error you get seems that lon/lat are exchanged or something like that. But the formulas you wrote look OK.

Exactly, the plot in my Notebook is quite relaxed and I did not took that into account. But this operations I wrote earlier represent the 'correct' traveled distance

**Reply by juanmanzanero** [reply]
Ah, one thing, also you have to flip the 'y', namely do y=-y. The circuit preprocessor has the Z normal pointing to the sky while the car simulator has the Z normal pointing to the road. I have to better document this, but the reasoning behind is that different references use these convention.

**Reply by fch-aa** [reply]
Thanks @juanmanzanero, that was it!

Blue = simulated
Gray = real

![LFG_Trackmap_simulation_vs_real](https://user-images.githubusercontent.com/21101725/166202368-2bcc3efa-e573-493f-849b-3bf40ee1c79d.png)

**Reply by juanmanzanero** [reply]
Well done! It looks super good
